numbat/src/interpreter.rs

```diff
@@ -137,137 +141,112 @@ mod tests {
-        );
+            speed(10 * meter, 2 * second)"), @"5 m/s");
     }
 
     #[test]
     fn power_operator() {
-        assert_evaluates_to_scalar("2^3", 2.0f64.powf(3.0));
-        assert_evaluates_to_scalar("-2^4", -(2.0f64.powf(4.0)));
-        assert_evaluates_to_scalar("2^(-3)", 2.0f64.powf(-3.0));
+        assert_snapshot!(evaluates("2^3"), @"8");
+        assert_snapshot!(evaluates("-2^4"), @"-16");
+        assert_snapshot!(evaluates("2^(-3)"), @"0.125");
     }
 
     #[test]
     fn multiline_input_yields_result_of_last_line() {
-        assert_evaluates_to_scalar("2\n3", 3.0);
+        assert_snapshot!(evaluates("2\n3"), @"3");
     }
 
     #[test]
     fn variable_definitions() {
-        assert_evaluates_to_scalar("let x = 2\nlet y = 3\nx + y", 2.0 + 3.0);
+        assert_snapshot!(evaluates("let x = 2\nlet y = 3\nx + y"), @"5");
     }
 
     #[test]
     fn function_definitions() {
-        assert_evaluates_to_scalar("fn f(x: Scalar) = 2 * x + 3\nf(5)", 2.0 * 5.0 + 3.0);
+        assert_snapshot!(evaluates(
+            "fn f(x: Scalar) = 2 * x + 3\nf(5)"),
+            @"13"
+        );
     }
 
     #[test]
     fn foreign_functions() {
-        assert_evaluates_to_scalar("sin(1)", 1.0f64.sin());
-        assert_evaluates_to_scalar("atan2(2 meter, 1 meter)", 2.0f64.atan2(1.0f64));
+        assert_snapshot!(evaluates("sin(1)"), @"0.841471");
+        assert_snapshot!(evaluates("atan2(2 meter, 1 meter)"), @"1.10715");
     }
 
     #[test]
     fn statistics_functions() {
-        assert_evaluates_to_scalar("mean(1, 1, 1, 0)", 0.75);
-        assert_evaluates_to(
-            "mean(1 m, 1 m, 1 m, 0 m)",
-            Quantity::new_f64(0.75, Unit::meter()),
+        assert_snapshot!(evaluates("mean(1, 1, 1, 0)"), @"0.75");
+        assert_snapshot!(evaluates(
+            "mean(1 m, 1 m, 1 m, 0 m)"),
+            @"0.75 m"
         );
-        assert_evaluates_to("mean(2 m, 100 cm)", Quantity::new_f64(1.5, Unit::meter()));
+        assert_snapshot!(evaluates("mean(2 m, 100 cm)"), @"1.5 m");
 
-        assert_evaluates_to_scalar("maximum(1, 2, 0, -3)", 2.0);
-        assert_evaluates_to(
-            "maximum(2 m, 0.1 km)",
-            Quantity::new_f64(100.0, Unit::meter()),
-        );
+        assert_snapshot!(evaluates("maximum(1, 2, 0, -3)"), @"2");
+        assert_snapshot!(evaluates(
+            "maximum(2 m, 0.1 km)"), @"100 m");
 
-        assert_evaluates_to_scalar("minimum(1, 2, 0, -3)", -3.0);
-        assert_evaluates_to(
-            "minimum(2 m, 150 cm)",
-            Quantity::new_f64(1.5, Unit::meter()),
-        );
+        assert_snapshot!(evaluates("minimum(1, 2, 0, -3)"), @"-3");
+        assert_snapshot!(evaluates(
+            "minimum(2 m, 150 cm)"), @"1.5 m");
     }
 
     #[test]
     fn division_by_zero_raises_runtime_error() {
-        assert_runtime_error("1/0", RuntimeError::DivisionByZero);
+        assert_snapshot!(evaluates("1/0"), @"DivisionByZero: `Division by zero`");
```

Comment:
This is somewhat similar. These tests now depend on the exact error message and might need to be updated whenever we change messages.

Reply:
Yeah, that’s similar, but here, I feel like it's 100% what we want, no?
The tests are easier to review, we can _read_ the error directly, and it shows us what will effectively be returned to the end user 🤔 

And in the end, when you change an error message, you can ensure straight away from the tests that it didn't have any side effect on other stuff and that it displays as you want. (this one is easy, but some error messages may contain variables and more complex formatting).

----

You have the last word, but personally, I think the only valuable information when testing errors as a binary is the final error message.
And for a library, it’s both the error message and the error type; that’s why I snapshotted both.

Reply:
> You have the last word, but personally, I think the only valuable information when testing errors as a binary is the final error message.

I think I agree.

> And for a library, it’s both the error message and the error type; that’s why I snapshotted both.

I mean.. I want both to be correct, but the important part about this test is that it asserts that divisions by zero *do raise* a runtime error (and the test doesn't care if the message is "Division by zero" or "Numerical error" or similar). Of course I also want the message to be correct, but I'm not sure if that needs a unit test. Ideally, I'd like to have that error message just once in my code base. But maybe I just need to adjust to the idea of snapshot tests. I'm okay with trying :smile: